<a href="https://colab.research.google.com/github/nedokormysh/Stepik_data_analysis_R/blob/week_1_data_prep/Stepik_data_analysis_R_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# mydata <- read.csv("https://stepik.org/media/attachments/lesson/11481/evals.csv", stringsAsFactors = T)

In [2]:
mydata <- mtcars
head(mydata)

mpg  cyl disp hp  drat wt    qsec  vs am gear carb
Mazda RX4         21.0 6   160  110 3.90 2.620 16.46 0  1  4    4   
Mazda RX4 Wag     21.0 6   160  110 3.90 2.875 17.02 0  1  4    4   
Datsun 710        22.8 4   108   93 3.85 2.320 18.61 1  1  4    1   
Hornet 4 Drive    21.4 6   258  110 3.08 3.215 19.44 1  0  3    1   
Hornet Sportabout 18.7 8   360  175 3.15 3.440 17.02 0  0  3    2   
Valiant           18.1 6   225  105 2.76 3.460 20.22 1  0  3    1

# 1_4_4

Создайте новую числовую переменную  new_var в данных mtcars, которая содержит единицы в строчках, если в машине не меньше четырёх карбюраторов (переменная "carb") или больше шести цилиндров (переменная "cyl"). В строчках, в которых условие не выполняется, должны стоять нули.

In [4]:
mtcars$new_var <- ifelse(mtcars$carb >= 4 | mtcars$cyl > 6, 1, 0)

# 1_4_5

В уже существующей переменной my_vector сохранен вектор из 50 чисел.

Решите задачу используя конструкцию:


if () {

} else {

}
Если среднее значение вектора my_vector больше 20, в переменную result сохраните "My mean is great",  если среднее значение my_vector меньше или равно 20 то в переменную result сохраните  строку "My mean is not so great".

In [5]:
my_vector <- c(20.67,23.34, 22.65, 17.11, 22.1,
               26.32, 20.39, 21.04, 23.78, 31.11,
               21.13, 22.44, 23.21, 27.02, 18.64,
               20.9, 20.77, 20.0, 21.29, 23.48,
               18.47, 25.02, 17.04, 30.97, 12.91,
               23.88, 32.95, 8.46 ,23.15, 21.05,
               20.63, 19.95, 17.38, 29.35, 24.43,
               23.66, 18.32, 30.13, 19.36, 19.67,
               24.23, 20.82, 18.21, 9.91 ,21.45,
               18.04, 18.31, 17.18, 10.99, 10.06)

In [7]:
if (mean(my_vector) > 20) {
    result <- ("My mean is great")
} else {
    result <- ("My mean is not so great")
}

[1] "My mean is great"


In [ ]:
if(mean(my_vector) > 20) {
  result <- ('My mean is great')
  print(result)
} else {
  result <- ('My mean is not so great')
  print(result)
}

# 1_4_6

В этой задаче от вас потребуется узнать некоторую информацию о типах данных в R самостоятельно! Встроенные в R данные AirPassengers - это новый для нас формат данных типа Time-Series. Изучите структуру этих данных, прежде чем начать решение задачи! Например напишите команды:

> ?AirPassengers # справка о данных
> str(AirPassengers) # структура данных
В встроенных в R данных AirPassengers хранится 144 значения (количество пассажиров в месяц) с 1949 по 1960 год. Данные Time-Series очень похожи на вектор по своей структуре, например мы можем обратиться к любому из 144 элементов используя уже знакомую нам индексацию AirPassengers[1] или AirPassengers[56].

Можно вообще перевести исходные данные в вектор при помощи команды as.vector(AirPassengers) и продолжить с ними работу как с вектором.

И так ваша задача создать переменную good_months и сохранить в нее число пассажиров только в тех месяцах, в которых это число больше, чем показатель в предыдущем месяце.  

Важный момент! В R оператор : для создания последовательности имеет приоритет над арифметическими действиями. Таким образом, если у вас есть переменная i, равная 10, и вы хотите создать вектор от 1 до i - 1, воспользуйтесь скобками, чтобы указать последовательность действий.

> i <- 10
> 1 : i - 1 # так мы создадим последовательность от 1 до 10, а потом вычтем единицу из каждого элемента
[1] 0 1 2 3 4 5 6 7 8 9

> 1 : (i - 1) # а вот так мы создадим последовательность от 1 до i - 1, то есть от 1 до 9.
[1] 1 2 3 4 5 6 7 8 9

In [8]:
str(AirPassengers)

 Time-Series [1:144] from 1949 to 1961: 112 118 132 129 121 135 148 148 136 119 ...


In [12]:
vector1 <- as.vector(AirPassengers)
vector2 <- c(vector1[length(vector1)],vector1[-length(vector1)])
diffs <- vector1-vector2
output = vector1[diffs>0]
good_months <- output
good_months

[1] 118 132 135 148 118 126 141 149 170 140 145 150 178 172 178 199 166 171 180
[20] 193 183 218 230 242 194 196 236 243 264 272 201 204 235 234 264 302 229 242
[39] 267 269 270 315 364 278 284 317 318 374 413 306 315 356 355 422 465 467 336
[58] 340 362 363 435 491 505 337 360 406 420 472 548 559 405 417 419 461 472 535
[77] 622 432

In [9]:
good_months <- c()
index <- 1
for (i in 2:length(AirPassengers)) {
	if (AirPassengers[i] > AirPassengers[i-1]){
		good_months[index] <- AirPassengers[i]
		index <- index + 1
		}
	}

In [10]:
good_months

[1] 118 132 135 148 118 126 141 149 170 140 145 150 178 172 178 199 166 171 180
[20] 193 183 218 230 242 194 196 236 243 264 272 201 204 235 234 264 302 229 242
[39] 267 269 270 315 364 278 284 317 318 374 413 306 315 356 355 422 465 467 336
[58] 340 362 363 435 491 505 337 360 406 420 472 548 559 405 417 419 461 472 535
[77] 622 432

In [13]:
AirPassengers[-1][AirPassengers[-1] > AirPassengers[-144]]

[1] 118 132 135 148 118 126 141 149 170 140 145 150 178 172 178 199 166 171 180
[20] 193 183 218 230 242 194 196 236 243 264 272 201 204 235 234 264 302 229 242
[39] 267 269 270 315 364 278 284 317 318 374 413 306 315 356 355 422 465 467 336
[58] 340 362 363 435 491 505 337 360 406 420 472 548 559 405 417 419 461 472 535
[77] 622 432

In [17]:
AirPassengers[0]

numeric(0)

# 1_3_7

3.8  Заключение
1.4 Элементы синтаксиса
1 из 8 шагов пройден
0 из 5 баллов  получено
Задачка для супер героев, повышенной сложности!

Для встроенных в R данных AirPassengers рассчитайте скользящее среднее с интервалом сглаживания равным 10. Напечатайте получившийся результат (первым значением в выводе должно быть среднее для элементов 1:10, во втором значении - среднее для элементов 2:11 и т.д., в последнем  - среднее для элементов 135 :144)

Все полученные значения средних сохраните в переменную moving_average.

Пример расчета для вектора из 11 элементов:









Соответственно, для наших данных из 144 наблюдений должно получиться 135 средних (первые и последние 4 средних):

129.8 129.0 129.0 127.3 ... 483.6 489.2 486.5 490.6

Важный момент! В R оператор : для создания последовательности имеет приоритет над арифметическими действиями. Таким образом, если у вас есть переменная i, равная 10, и вы хотите создать вектор от 1 до i - 1, воспользуйтесь скобками, чтобы указать последовательность действий.
> i <- 10
> 1 : i - 1 # так мы создадим последовательность от 1 до 10, а потом вычтем единицу из каждого элемента
[1] 0 1 2 3 4 5 6 7 8 9

> 1 : (i - 1) # а вот так мы создадим последовательность от 1 до i - 1, то есть от 1 до 9.
[1] 1 2 3 4 5 6 7 8 9

Если вам потребуется создать вектор moving_average заранее, то есть несколько способов сделать это:
1. самый простой, но не очень правильный вариант - создать пустой вектор
moving_average <- c()

2. можно сразу создать вектор определенной длины и определенного типа:
moving_average <- numeric(135)
Такой вариант является более предпочтительным. Почему? Узнаем во второй части курса!)

А для тем, кто уже выбрал путь воина и не хочет использовать цикл - советую познакомиться с функцией cumsum. Подсказка: если у нас есть два вектора одинаковой длинны, то если из одного вектора вычесть второй вектор, мы найдем разность для первых элементов векторов, затем для вторых и т.д.


> x <- c(2, 4, 7)

> y <- c(2, 3, 5)

> x - y

[1] 0 1 2